src/dns.rs

```diff
@@ -0,0 +1,510 @@
+use crate::runtime::Runtime;
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::error;
+use std::fmt;
+use std::io;
+use std::net::AddrParseError;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver {
+    resolver: Option<AsyncResolver>,
+}
+
+impl DnsResolver {
+    pub fn new<'a, R: Into<Option<&'a mut Runtime>>>(
```

Comment:
I think this can just return a future to let the call spawn or not. This means when the executor is passed down we don't have to change this code and it can just spawn it in a more clear manner.